File: Yixuan/16_01_17/last_position_of_target.py
class Solution:
    # @param {int[]} A an integer array sorted in ascending order
    # @param {int} target an integer
    # @return {int} an integer
    # @link: http://www.lintcode.com/en/problem/last-position-of-target/
    # @author: Egbert Li
    def lastPosition(self, A, target):
        if len(A) == 0:
            return -1
        start = 0
        end = len(A) - 1
        while start + 1< end:
            mid = int(start + (end - start) / 2 )
            if target < A[mid]:
                end = mid
            else:
                start  = mid
        if target == A[end]:
            return end
        elif target == A[start]:
            return start
        else:
            return -1

File: Yixuan/16_01_17/last_position_of_target.cpp
class Solution {
public:
    /**
     * @param A an integer array sorted in ascending order
     * @param target an integer
     * @return an integer
     * @link: http://www.lintcode.com/en/problem/last-position-of-target/
     * @author: Egbert Li
     */
    int lastPosition(vector<int>& A, int target) {
      int length = A.size();
      if (length == 0) {
         return -1;
      }
      int start = 0;
      int end = A.size() - 1;
      while (start + 1 < end) {
         int mid = start + (end - start) / 2;
         if (target < A[mid]) {
            end = mid;
         } else {
            start = mid;
         }
      }
      if (target == A[end]) {
         return end;
      } else if (target == A[start]) {
         return start;
      } else {
         return -1;
      }
    }
};
